In [3]:
import pandas as pd

df = pd.read_csv("data/book400k-500k.csv")

df = df.sample(n=50000, random_state=42)

df.to_csv("book_sample.csv", index=False)

print("Sampling complete!")

Sampling complete!


In [4]:
print(df.columns)

Index(['PublishYear', 'Rating', 'RatingDistTotal', 'ISBN', 'RatingDist1',
       'Publisher', 'PublishMonth', 'Id', 'Name', 'Authors', 'RatingDist5',
       'RatingDist4', 'PublishDay', 'RatingDist2', 'pagesNumber',
       'RatingDist3', 'CountsOfReview', 'Language'],
      dtype='str')


In [5]:
df_sorted = df.sort_values("Id").reset_index(drop=True)
search_list = df_sorted["Id"].tolist()

print(f"Data sorted, ready to search through {len(search_list)} records.")

Data sorted, ready to search through 50000 records.


In [6]:
def linear_search(data, target):
    for i in range(len(data)):
        if data[i] == target:
            return i  # Found
    return -1  # Not found

In [7]:
#Testing

# Pick a random ID from your list to search for
target_id = search_list[45000] 

# Run your search
result = linear_search(search_list, target_id)

print(f"Found the book at index: {result}")

Found the book at index: 45000


In [8]:
#Linear search
def linear_search(data, target):
    for i in range(len(data)):
        if data[i] == target:
            return i
    return -1

In [9]:
#Binary search
def binary_search(data, target):
    low, high = 0, len(data) - 1
    while low <= high:
        mid = (low + high) // 2
        if data[mid] == target: return mid
        elif data[mid] < target: low = mid + 1
        else: high = mid - 1
    return -1

In [10]:
#Jump search
import math
def jump_search(data, target):
    n = len(data)
    step = int(math.sqrt(n))
    prev = 0
    while data[min(step, n)-1] < target:
        prev = step
        step += int(math.sqrt(n))
        if prev >= n: return -1
    for i in range(prev, min(step, n)):
        if data[i] == target: return i
    return -1

In [11]:
#Interpolation search
def interpolation_search(data, target):
    low, high = 0, len(data) - 1
    while low <= high and target >= data[low] and target <= data[high]:
        if low == high:
            if data[low] == target: return low
            return -1
        pos = low + int(((float(high - low) / (data[high] - data[low])) * (target - data[low])))
        if data[pos] == target: return pos
        if data[pos] < target: low = pos + 1
        else: high = pos - 1
    return -1

In [ ]:
import time

def benchmark_search(search_func, data, target):
    times = []
    for _ in range(3):
        start = time.perf_counter()
        search_func(data, target)
        end = time.perf_counter()
        times.append((end - start) * 1000) # Convert to ms
    return sum(times) / len(times)

# Test sizes required
sizes = [1000, 10000, 50000]
algorithms = [linear_search, binary_search, jump_search, interpolation_search]

for n in sizes:
    subset = search_list[:n]
    target = subset[-1] # Target the last item for worst-case timing
    print(f"\nResults for n={n}:")
    for alg in algorithms:
        avg_time = benchmark_search(alg, subset, target)
        print(f"{alg.__name__}: {avg_time:.6f} ms")


Results for n=1000:
linear_search: 0.043922 ms
binary_search: 0.003815 ms
jump_search: 0.027535 ms
interpolation_search: 0.002431 ms

Results for n=10000:
linear_search: 0.697667 ms
binary_search: 0.003765 ms
jump_search: 0.065923 ms
interpolation_search: 0.001632 ms

Results for n=50000:
linear_search: 2.946851 ms
binary_search: 0.002141 ms
jump_search: 0.066747 ms
interpolation_search: 0.000948 ms


**Specific Input:** A dataset with a massive "outlier" (e.g., [1, 2, 3, 4, 1000000]).

**Explanation:** Interpolation search performs poorly here because the large gap in values makes its position "guess" highly inaccurate, forcing it to behave more like a slow linear search.




**Performance Comparison**

n = 1,000: Interpolation Search wins at 0.002431 ms.

n = 10,000: Interpolation Search wins at 0.001632 ms.

n = 50,000: Interpolation Search wins at 0.000948 ms.

In [13]:
#Bad data set, that performs worse
bad_data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 1000000]
target = 9

binary_time = benchmark_search(binary_search, bad_data, target)
interp_time = benchmark_search(interpolation_search, bad_data, target)

print(f"Binary Search: {binary_time:.6f} ms")
print(f"Interpolation Search: {interp_time:.6f} ms")

Binary Search: 0.001426 ms
Interpolation Search: 0.006806 ms
